## Importing Libraries

 - Importing necessary libraries for the analysis.

In [1]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

 - Pandas is used loading and viewing our dataset, also to process, clean and manipulate dataset.
 - Seaborn and Matplotlib to plot and visualize dataset.

## Load Dataset

 - Loading our dataset, here we use the 'pd.read_csv' function from pandas to locate and read it. In Jupyter notebook, because this ipynb file has the same location with the csv, it can be automatically run.
 - Here i give aditional option  to display all the columns, by default, pandas only display several columns and not all of them.
 - Displaying the top 6 rows of dataset.

In [2]:
df = pd.read_csv('Telcom.csv')
pd.set_option('display.max_columns', None)
df.head(6)

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   
5  9305-CDSKC  Female              0      No         No       8          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   
5               Yes     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   
5              Yes          No         Yes             Yes  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges TotalCharges  \
0              Yes           Electronic check           29.85        29.85   
1               No               Mailed check           56.95       1889.5   
2              Yes               Mailed check           53.85       108.15   
3               No  Bank transfer (automatic)           42.30      1840.75   
4              Yes           Electronic check           70.70       151.65   
5              Yes           Electronic check           99.65        820.5   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes  
5   Yes

## Data Profiling

 - The Next step we want to do, after we get the data and previewing it. We want to do a data profiling.
 - Data Profiling may include, determining how many columns and rows, what is the index of each of the column, what is the dimension of dataset, et cetera.


In [3]:
print("Dataset dimension :", df.ndim)
print("Dataset shape :", df.shape)
print("Dataset size :", df.size)

Dataset dimension : 2
Dataset shape : (7043, 21)
Dataset size : 147903


 - Here we can infer that dataset has 2 dimensions, comprises of 7043 observations (rows) and 21 charateristics (columns), and multiply those two numbers, we have a total 147903 number of elements of our dataset.
 - Now we know how many columns there are, but what each of the columns assigned for (indexed)?

In [4]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

 - By using the above function we know the index of each column. Such as, customerID, Partner, Contract, and so on.
 - Next we want to determine what type of each column falls into.

In [5]:
df.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

 - Dataset is dominated by object / string Dtype with total of 18 variables, while the rest 3 variables are mix of integer and float.

## Dataset Brief Explanation

 - The dataset is dominated mostly by categorical variables. As of now, the only numerical variables consist of only 5, while the rest are categorical.
 - There are couple of grouping of different types of measurement and understanding them better will help us learn the data aswell.
 - Subscriber's information, the brief information about the subcriber. For example: CustomerID, gender, SeniorCitizen, Partner, Dependents. and tenure period.
 - The type of service the company provides- PhoneService, MultipleLines, InternetService.
 - The Additional features / bundle added into the services they provide - OnlineSecurity, OnlineBackup, DeviceProtection, TechSupport, StreamingTV and  StreamingMovies.
 - Subscribers setup and annual bill for the services - Contract, PaperlessBilling, PaymenMethod.MonthlyCharges and totalCharges.
 
 
 - customerID, number assigned and is unique to each of the subscriber.
 - gender, specifies the gender of the subscriber.
 - SeniorCitizen, specifies whether a subscriber is an elderly or not.
 - Partner, specifies whether a subscriber has married (has a partner).
 - Dependent, specifies whether a subscriber pays the bill on their own or need someone to help pay the bill.
 - tenure, specifies how long a subscriber is using the company's service.
 - PhoneService, specifies whether a subscriber is using the phone service provided by the company.
 - MultipleLines, specifies whether a subscriber is using a combination of services at once, provided by the company.
 - InternetServices, types of internet used by each of the subscriber.
 - OnlineSecurity, specifies whether a subsriber is using the additional services of online / surfing security.
 - OnlineBackup, specifies wheter a subscriber is using the additional service of online data backup.
 - DeviceProtection, specifies whther a subscriber is using the additional service of device / router protection.
 - TechSupport, specifies whether a subscriber need the help of a technician should they come across a problem.
 - StreamingTV, specifies whether a subscriber want streaming tv services added to their packages.
 - StreamingMovies, specifies whether a subscriber want streaming movie services added to their package.
 - Contract, specifies how a subscriber want to prolong / extent using the company's service.
 - PaperlessBilling, specifies how the subscriber's bill is sent to them.
 - PaymenMethod, which platform a subscriber is paying their annual bill.
 - MonthlyCharges, the amount of money to be paid per month.
 - TotalCHarges, the grand total amount of money to be paid.
 - The last is Churn, refers to a condition where a subscribers ever cancelled or stop using the service.

## Checking for Missing Values

 - Check to see if there's any missing value from our dataset.
 - Here we are going to make  a new dataframe consisting of 'Filling Factor', which counts how much is the missing value in dataset for a given columns. It is set in percentage.

In [9]:
missing_data = df.isna().sum(axis=0).reset_index()
missing_data.columns = ['Variable','Missing Value']
blank_data = df.apply(lambda x : x == ' ').sum(axis=0).reset_index()
blank_data.columns = ['Variable','Blank Value']
merged_data = pd.merge(missing_data, blank_data, on=["Variable"])
merged_data['Filling Factor %'] = (df.shape[0]-merged_data['Missing Value'] + df.shape[0]-merged_data['Blank Value'] )/df.shape[0]*50
merged_data.sort_values('Filling Factor %').reset_index(drop=False)

index          Variable  Missing Value  Blank Value  Filling Factor %
0      19      TotalCharges              0           11         99.921908
1       0        customerID              0            0        100.000000
2      18    MonthlyCharges              0            0        100.000000
3      17     PaymentMethod              0            0        100.000000
4      16  PaperlessBilling              0            0        100.000000
5      15          Contract              0            0        100.000000
6      14   StreamingMovies              0            0        100.000000
7      13       StreamingTV              0            0        100.000000
8      12       TechSupport              0            0        100.000000
9      11  DeviceProtection              0            0        100.000000
10     10      OnlineBackup              0            0        100.000000
11      8   InternetService              0            0        100.000000
12      7     MultipleLines              0            0        100.000000
13      6      PhoneService              0            0        100.000000
14      5            tenure              0            0        100.000000
15      4        Dependents              0            0        100.000000
16      3           Partner              0            0        100.000000
17      2     SeniorCitizen              0            0        100.000000
18      1            gender              0            0        100.000000
19      9    OnlineSecurity              0            0        100.000000
20     20             Churn              0            0        100.000000

 - As we can see, dataset contains no missing values.
 - However, we may want to double-check if there's legit no missing value in dataset.
 - We may want to check aside of missing values, we have to consider a blank space (' ') as a missing value as well.
 - We apply lambda function to do so.

In [ ]:
df.apply(lambda x : x == ' ').sum()

 - We can see it very clearly in TotalCharges there are 11 blank spaces.
 - There are defiintely missing values in dataset.
 - Locating the datas with the blank spaces in dataset, using filter and then locating them with df.loc. 

In [ ]:
filt = (df['TotalCharges'] == (' '))
df.loc[filt]

 - We can now confirm there are 11 blank spaces in TotalCharges.
 - The data needs to be cleaned of missing value, proceed to data cleaning.

## Data Cleaning

 - This process can be done by simply filling the blank space in TotaCharges variable by 0. Using the similar lambda method just like the above, we can replace the blank spaces by the value of 0.
 - However, first we have to confirm the columns type of TotalCharges. Using df.types, we can see what types that variable falls into.

In [ ]:
df.dtypes

 - The variable is still in object / string. Have to change it to either integer or float before. Now we know from dataset preview that some of the values of this variable are decimal, then float is the choice.
 - Next, is we replace the blank spaces by 0.0, because we choose float as our type for TotalCharges.
 - Using df.replace() to change blank spaces (' ') to ('0.0'), and then locate the replaced datas by df.loc.

In [ ]:
df['TotalCharges'] = df['TotalCharges'].replace([' '],['0.0'])
df.loc[(df['TotalCharges'] == '0.0')]

 - We can see now that the initial blank spaces of the variable have now been replaced by the 0.0's. 
 - This time we can properly change TotalCharges variable into its proper type, float.

In [ ]:
df['TotalCharges'] = df['TotalCharges'].astype(float)

 - Again, confirming that the variable's type has been changed using df.dtypes.

In [ ]:
df.dtypes

 - We might want to further look into SeniorCitizen as it has a value 0 but instead its variable type is object / string and not integer. 
 - Check to see how many unique values for SeniorCitizen variable, using df.unique and df.value_counts

In [ ]:
print('Unique value(s) of SenioCitizen : ', df['SeniorCitizen'].unique())
print(df['SeniorCitizen'].value_counts())

 - The SeniorCitizen should be of categorical data, as it only has two values '0' and '1'.
 - Next is changing this variable into a string / object type value, also replacing the both value of '0' and '1' to 'No' and 'Yes' 
 - Using the same df.replace method as the above to replace the values of '0' and '1' to 'No' and 'Yes' respectively.

In [ ]:
df['SeniorCitizen'] = df['SeniorCitizen'].astype(object)
df['SeniorCitizen'] = df['SeniorCitizen'].replace([0,1],['No','Yes'])
df.head(6)

In [ ]:
df['SeniorCitizen'].value_counts()

 - As we can see, both of the 0 and 1 have been replaced for SeniorCitizen variable.
 - Next is outlers checking and handling them.

## Outliers Checking 

 - Checking and determining data outliers for each of the SeniorCitizen, tenure, MonthlyCharges and TotalCharges variable using IQR, 25th %tile and 75th %tile calculation. 

In [ ]:
###

#tenure
Q1_2 = df['tenure'].quantile(0.25)
Q3_2 = df['tenure'].quantile(0.75)
IQR_2 = Q3_2 - Q1_2
LW_2 = Q1_2 - 1.5*IQR_2
UB_2 = Q3_2 + 1.5*IQR_2
OL_2 = ((df['tenure'] < LW_2) | (df['tenure'] > UB_2)).sum()

###

#MonthlyCharges
Q1_3 = df['MonthlyCharges'].quantile(0.25)
Q3_3 = df['MonthlyCharges'].quantile(0.75)
IQR_3 = Q3_3 - Q1_3
LW_3 = Q1_3 - 1.5*IQR_3
UB_3 = Q3_3 + 1.5*IQR_3
OL_3 = ((df['MonthlyCharges'] < LW_3) | (df['MonthlyCharges'] > UB_3)).sum()

###

#TotalCharges
Q1_4 = df['TotalCharges'].quantile(0.25)
Q3_4 = df['TotalCharges'].quantile(0.75)
IQR_4 = Q3_4 - Q1_4
LW_4 = Q1_4 - 1.5*IQR_4
UB_4 = Q3_4 + 1.5*IQR_4
OL_4 = ((df['TotalCharges'] < LW_4) | (df['TotalCharges'] > UB_4)).sum()

###

# Creating Outliers Dataframe 
df_outliers = pd.DataFrame({'Variables' : ['tenure','MonthlyCharges','TotalCharges'],
                           '25th %tile' : [Q1_2,Q1_3,Q1_4],
                           '75th %tile' : [Q3_2,Q3_3,Q3_4],
                           'IQR' : [IQR_2,IQR_3,IQR_4],
                           'Lower Bound' : [LW_2,LW_3,LW_4],
                           'Upper Bound' : [UB_2,UB_3,UB_4],
                           'Outliers count' : [OL_2,OL_3,OL_4]})
df_outliers

 - No outliers found on dataset. All of the data meets the upper and lower boundaries for each tenure, MonthlyCharges and TotalCharges variable.
 - Running a boxplot for above variables, double-checking to see if there's any outliers.
 - Using seaborn package, we can plot a box plot diagram.

In [ ]:
ax1 = sns.boxplot(y='MonthlyCharges', data=df, color='green', orient='v')
sns.set_style('whitegrid')
ax1.set_title("Boxplot of Charges per Month", fontsize = 15)
ax1.set_xlabel("")
ax1.set_ylabel("")
plt.show()


ax2 = sns.boxplot(y='TotalCharges', data=df, color='green', orient='v')
sns.set_style('whitegrid')
ax2.set_title("Boxplot of Total Charges", fontsize = 15)
ax2.set_xlabel("")
ax2.set_ylabel("")
plt.show()


ax3 = sns.boxplot(y='tenure', data=df, color='green', orient='v')
sns.set_style('whitegrid')
ax3.set_title("Boxplot Tenure Period", fontsize = 15)
ax3.set_xlabel("")
ax3.set_ylabel("")
plt.show()

 - There are no values which exceeds the upper and lower boundaries of the boxplot in those 3 variables.
 - We can assurely confirm there are no outliers in dataset.

## Feature Engineering

 - Feature Engineering is a technique to create new features from the existing data that could help to gain more insight into the data. 
 - Lets say we to make a new feature which enables us to group values from  tenure variable into 5 different categories. 
 - For this, we are using pandas qcut() function to group them. Say we name the feature 'tenurecategories'.

In [ ]:
lbs = ['Very Short','Short','Middle','Long','Very Long']
df['tenurecategories'] = pd.qcut(df['tenure'], q=5, labels=lbs)
df[['tenure','tenurecategories']].head(16)

 - As we can see we have assigned a new feature which groups tenure variable into 5 different categories using pandas.
 - Now perhaps we may want to add a new feature which enables us to count how many values for each different categories.

In [ ]:
pd.DataFrame(df['tenurecategories'].value_counts())

## Data Normalization

 - Normalization is a rescaling of the data from the original range so that all values are within the new range of 0 and 1.
 - Here we want to Normalize tenure, MonthlyCharges and TotalCharges variable from our dataset.
 - We can normalize dataset using the scikit-learn object MinMaxScaler.

In [ ]:
from sklearn.preprocessing import MinMaxScaler
data1 = df[['tenure','MonthlyCharges','TotalCharges']]

# Define MinMaxScaler
scaler1 = MinMaxScaler()

# Transform data
scaled1 = scaler1.fit_transform(data1)

# Turning into Dataframe
dfnormal = pd.DataFrame(scaled1)
dfnormal.columns = ['tenure_nrm', 'MonthlyCharges_nrm','TotalCharges_nrm']
dfnormal

## Data Standarization

 - Standardizing a dataset involves rescaling the distribution of values so that the mean of observed values is 0 and the standard deviation is 1.
 - We can standardize our dataset using the scikit-learn object StandardScaler.

In [ ]:
from sklearn.preprocessing import StandardScaler
data2 = dfnormal

# Define StandardScaler
scaler2 = StandardScaler()

# Transform data
scaled2 = scaler2.fit_transform(data2)

# Turning into Dataframe
dfstd = pd.DataFrame(scaled2)
dfstd.columns = ['tenure_std', 'MonthlyCharges_std','TotalCharges_std']
dfstd

## Comparison

 - Now we want to compare the descriptive statistics attributes for each of df, dfnormal and dfstd.

In [ ]:
print(df.describe())
print(dfnormal.describe())
print(dfstd.describe())

## Categorical Data Encoding

 - Implied above, we know that much of the dataset's variable are predominantly in string type. As such, it is much more useful and easier if we group and classify them using this method.
 - However, we may want to drop the customerID variable, as the value of this variables are unique to each customer (7043 different values) then in order to simplify the process we will drop it. 

In [ ]:
df.drop('customerID', axis=1, inplace=True)
df_dummy = pd.get_dummies(df)
df_dummy.head()